### PR TITLE
Add chat support to text generation pipeline

### PR DESCRIPTION
**Pull request by Rocketknight1**
This PR modifies the text generation pipeline to support chats. It does this by inspecting the inputs - if they look like strings, it uses the original causal LM pipeline, and if they look like lists of message dicts, it applies a chat template instead before proceeding with generation.

Most changes are in the preprocessing/postprocessing - the actual generation itself is largely unchanged.

TODO:
- [x] Expand tests to cover other edge cases
- [x] Confirm the return format we want for this - just the model response, or the entire chat?
- [x] ~Add KV cache support, as this is important for performant multi-turn chat~
- [x] Deprecate `ConversationalPipeline` and update the chat template docs to refer to this instead?

cc @ArthurZucker @gante @LysandreJik

### CODE REVIEW

**Review by julien-c — COMMENTED** [review]
looks neat from a (very) superficial glance

I think this will be quite useful!

**Comment by Rocketknight1** [comment]
@julien-c Done! This PR now adds a `DeprecationWarning` to `ConversationalPipeline`. I also updated the chat template docs for the new pipeline.

**Comment by julien-c** [comment]
very nice!

File: src/transformers/pipelines/conversational.py [hunk 1]
```diff
@@ -232,6 +233,10 @@ class ConversationalPipeline(Pipeline):
     """
 
     def __init__(self, *args, **kwargs):
+        warnings.warn(
+            "`ConversationalPipeline` is now deprecated, and the functionality has been moved to the standard `text-generation` pipeline, which now accepts lists of message dicts as well as strings. This class will be removed in a future release.",
```
Comment (gante) [inline src/transformers/pipelines/conversational.py:237]: I'd add a specific version in which the class will be deprecated (v4.40, i.e. two minor releases after next minor release?)

File: src/transformers/pipelines/text_generation.py [hunk 2]
```diff
@@ -20,11 +20,24 @@ class ReturnType(enum.Enum):
     FULL_TEXT = 2
 
 
+class Chat:
+    """This class is intended to just be used internally in this pipeline and not exposed to users. We convert chats
+    to this format because the rest of the pipeline code tends to assume that lists of messages are
+    actually a batch of samples rather than messages in the same conversation."""
+
+    def __init__(self, messages):
```
Comment (gante) [inline src/transformers/pipelines/text_generation.py:28]:
```suggestion
    def __init__(self, messages: dict):
```
nit: helps understanding that a dict is expected

File: src/transformers/pipelines/text_generation.py [hunk 3]
```diff
@@ -216,7 +230,15 @@ def __call__(self, text_inputs, **kwargs):
             - **generated_token_ids** (`torch.Tensor` or `tf.Tensor`, present when `return_tensors=True`) -- The token
               ids of the generated text.
         """
-        return super().__call__(text_inputs, **kwargs)
+        if isinstance(text_inputs, (list, tuple)) and isinstance(text_inputs[0], (list, tuple, dict)):
+            # We have one or more prompts in list-of-dicts format, so this is chat mode
+            if isinstance(text_inputs[0], dict):
+                return super().__call__(Chat(text_inputs), **kwargs)
+            else:
+                chats = [Chat(chat) for chat in text_inputs]  # 🐈 🐈 🐈
```
Comment (gante) [inline src/transformers/pipelines/text_generation.py:238]: best comment 😂

**Review by gante — APPROVED** [review]
Nice! Thank for adding this 🐈

File: src/transformers/pipelines/text_generation.py (hunk 2 above)
Comment (Rocketknight1) [reply src/transformers/pipelines/text_generation.py]: Done!

File: src/transformers/pipelines/conversational.py (hunk 1 above)
Comment (Rocketknight1) [reply src/transformers/pipelines/conversational.py]: Done! I picked 4.42 to give people more time.

**Comment by gante** [comment]
IMO it looks good to me

**Comment by Rocketknight1** [comment]
Cool!

**Comment by Rocketknight1** [comment]
In that case, I think we're ready for final review (cc @amyeroberts)  - I'm leaving the KV cache to another PR.

**Comment by Rocketknight1** [comment]
cc @LysandreJik @julien-c as well if there's anything else you want me to add before we merge this!

File: src/transformers/pipelines/text_generation.py [hunk 4]
```diff
@@ -229,14 +251,25 @@ def preprocess(
...
+            )
+        else:
+            inputs = self.tokenizer(
+                prefix + prompt_text,
+                return_tensors=self.framework,
+                truncation=truncation,
+                padding=padding,
+                max_length=max_length,
+                add_special_tokens=add_special_tokens,
+            )
```
Comment (amyeroberts) [inline src/transformers/pipelines/text_generation.py:272]: nit - same here (+ my own personal preference for `return_tensors` being the last option passed)

```suggestion
            inputs = self.tokenizer.apply_chat_template(
                prompt_text.messages,
                truncation=truncation,
                padding=padding,
                max_length=max_length,
                add_generation_prompt=True,
                return_dict=True,
                return_tensors=self.framework,
            )
        else:
            inputs = self.tokenizer(
                prefix + prompt_text,
                truncation=truncation,
                padding=padding,
                max_length=max_length,
                add_special_tokens=add_special_tokens,
                return_tensors=self.framework,
            )
```

File: src/transformers/pipelines/text_generation.py [hunk 5]
```diff
@@ -216,7 +230,15 @@ def __call__(self, text_inputs, **kwargs):
             - **generated_token_ids** (`torch.Tensor` or `tf.Tensor`, present when `return_tensors=True`) -- The token
               ids of the generated text.
         """
-        return super().__call__(text_inputs, **kwargs)
+        if isinstance(text_inputs, (list, tuple)) and isinstance(text_inputs[0], (list, tuple, dict)):
```
Comment (amyeroberts) [inline src/transformers/pipelines/text_generation.py:233]: Just to make sure - is it not possible for someone to pass this to the pipeline:

```
# Pass a list-of-list-of-strings
generator([["this is a dog"], ["this is a code example"], ["banana for scale"]])
```

**Review by amyeroberts — APPROVED** [review]
Beautiful - thanks for adding this support!

File: src/transformers/pipelines/text_generation.py (hunk 5 above)
Comment (Rocketknight1) [reply src/transformers/pipelines/text_generation.py:233]: I tried that on `main` - it just results in a `TypeError: can only concatenate str (not "list") to str`. The existing pipeline will only accept either a single string or a non-nested list/tuple of strings, so I don't think this check makes a mistake for any valid inputs!

**Pull request merged**